File: rest/src/main/java/io/sparkled/rest/service/stage/StageRestServiceHandler.java
package io.sparkled.rest.service.stage;

import com.google.inject.persist.Transactional;
import io.sparkled.model.entity.Stage;
import io.sparkled.model.entity.StageProp;
import io.sparkled.persistence.stage.StagePersistenceService;
import io.sparkled.rest.response.IdResponse;
import io.sparkled.rest.service.RestServiceHandler;
import io.sparkled.viewmodel.stage.StageViewModel;
import io.sparkled.viewmodel.stage.StageViewModelConverter;
import io.sparkled.viewmodel.stage.prop.StagePropViewModel;
import io.sparkled.viewmodel.stage.prop.StagePropViewModelConverter;

import javax.inject.Inject;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

class StageRestServiceHandler extends RestServiceHandler {

    private final StagePersistenceService stagePersistenceService;
    private final StageViewModelConverter stageViewModelConverter;
    private final StagePropViewModelConverter stagePropViewModelConverter;

    @Inject
    public StageRestServiceHandler(StagePersistenceService stagePersistenceService,
                                   StageViewModelConverter stageViewModelConverter,
                                   StagePropViewModelConverter stagePropViewModelConverter) {
        this.stagePersistenceService = stagePersistenceService;
        this.stageViewModelConverter = stageViewModelConverter;
        this.stagePropViewModelConverter = stagePropViewModelConverter;
    }

    @Transactional
    Response createStage(StageViewModel stageViewModel) {
        Stage stage = stageViewModelConverter.toModel(stageViewModel);
        stage = stagePersistenceService.createStage(stage);
        return getJsonResponse(new IdResponse(stage.getId()));
    }

    Response getAllStages() {
        List<Stage> stages = stagePersistenceService.getAllStages();
        return getJsonResponse(stages);
    }

    Response getStage(int stageId) {
        Optional<Stage> stageOptional = stagePersistenceService.getStageById(stageId);

        if (stageOptional.isPresent()) {
            Stage stage = stageOptional.get();
            StageViewModel viewModel = stageViewModelConverter.toViewModel(stage);

            List<StagePropViewModel> stageProps = stagePersistenceService
                    .getStagePropsByStageId(stageId)
                    .stream()
                    .map(stagePropViewModelConverter::toViewModel)
                    .collect(Collectors.toList());
            viewModel.setStageProps(stageProps);

            return getJsonResponse(viewModel);
        }

        return getResponse(Response.Status.NOT_FOUND);
    }

    @Transactional
    Response updateStage(int id, StageViewModel stageViewModel) {
        stageViewModel.setId(id); // Prevent client-side ID tampering.

        Stage stage = stageViewModelConverter.toModel(stageViewModel);
        List<StageProp> stageProps = stageViewModel.getStageProps()
                .stream()
                .map(stagePropViewModelConverter::toModel)
                .map(ps -> ps.setStageId(id))
                .collect(Collectors.toList());

        stagePersistenceService.saveStage(stage, stageProps);
        return getResponse(Response.Status.OK);
    }

    @Transactional
    Response deleteStage(int id) {
        stagePersistenceService.deleteStage(id);
        return getResponse(Response.Status.OK);
    }
}

File: rest/src/main/kotlin/io/sparkled/rest/service/stage/StageRestServiceHandler.kt
package io.sparkled.rest.service.stage

import io.sparkled.persistence.stage.StagePersistenceService
import io.sparkled.persistence.transaction.Transaction
import io.sparkled.rest.response.IdResponse
import io.sparkled.rest.service.RestServiceHandler
import io.sparkled.viewmodel.stage.StageViewModel
import io.sparkled.viewmodel.stage.StageViewModelConverter
import io.sparkled.viewmodel.stage.prop.StagePropViewModelConverter
import io.sparkled.viewmodel.stage.search.StageSearchViewModelConverter
import javax.inject.Inject
import javax.ws.rs.core.Response

class StageRestServiceHandler
@Inject constructor(
    private val transaction: Transaction,
    private val stagePersistenceService: StagePersistenceService,
    private val stageViewModelConverter: StageViewModelConverter,
    private val stageSearchViewModelConverter: StageSearchViewModelConverter,
    private val stagePropViewModelConverter: StagePropViewModelConverter
) : RestServiceHandler() {

    fun createStage(stageViewModel: StageViewModel): Response {
        return transaction.of {
            var stage = stageViewModelConverter.toModel(stageViewModel)
            stage = stagePersistenceService.createStage(stage)
            return@of respondOk(IdResponse(stage.getId()!!))
        }
    }

    fun getAllStages(): Response {
        val stages = stagePersistenceService.getAllStages()
        val viewModels = stageSearchViewModelConverter.toViewModels(stages)
        return respondOk(viewModels)
    }

    fun getStage(stageId: Int): Response {
        val stage = stagePersistenceService.getStageById(stageId)

        if (stage != null) {
            val viewModel = stageViewModelConverter.toViewModel(stage)

            val stageProps = stagePersistenceService
                .getStagePropsByStageId(stageId)
                .asSequence()
                .map(stagePropViewModelConverter::toViewModel)
                .toList()
            viewModel.setStageProps(stageProps)

            return respondOk(viewModel)
        }

        return respond(Response.Status.NOT_FOUND, "Stage not found.")
    }

    fun updateStage(id: Int, stageViewModel: StageViewModel): Response {
        return transaction.of {
            stageViewModel.setId(id) // Prevent client-side ID tampering.

            val stage = stageViewModelConverter.toModel(stageViewModel)
            val stageProps = stageViewModel.getStageProps()
                .asSequence()
                .map(stagePropViewModelConverter::toModel)
                .map { it.setStageId(id) }
                .toList()

            stagePersistenceService.saveStage(stage, stageProps)
            return@of respondOk()
        }
    }

    fun deleteStage(id: Int): Response {
        return transaction.of {
            stagePersistenceService.deleteStage(id)
            return@of respondOk()
        }
    }
}